engine/table/src/main/java/io/deephaven/engine/table/impl/TableUpdateMode.java

```diff
@@ -31,4 +41,43 @@ public boolean removeAllowed() {
                 return false;
         }
     }
+
+    /**
+     * Returns the most permissive mode from the given stream of modes. Permissiveness is ranked as follows (from most
+     * to least permissive):
+     * <ul>
+     * <li>{@link #ADD_REMOVE}</li>
+     * <li>{@link #ADD_ONLY}</li>
+     * <li>{@link #APPEND_ONLY}</li>
+     * <li>{@link #STATIC}</li>
+     * </ul>
+     *
+     * @param modes a stream of modes
+     * @return the most permissive mode encountered in the stream
+     */
+    public static TableUpdateMode mostPermissiveMode(Stream<TableUpdateMode> modes) {
```

Comment:
As-is, this could be simplified to sorting. I guess it could become more complicated if we introduce capabilities that don't "nest" in the future.